server/src/main/java/org/elasticsearch/index/seqno/ReplicationTracker.java

```diff
@@ -384,21 +384,24 @@ private RetentionLease innerAddRetentionLease(String id, long retainingSequenceN
      * @param source                  the source of the retention lease
      * @return the renewed retention lease
      * @throws RetentionLeaseNotFoundException if the specified retention lease does not exist
+     * @throws IllegalArgumentException        if the new retaining sequence number is lower than
+     *                                         the retaining sequence number of the current retention lease.
      */
     public synchronized RetentionLease renewRetentionLease(final String id, final long retainingSequenceNumber, final String source) {
         assert primaryMode;
-        if (retentionLeases.contains(id) == false) {
+        final RetentionLease existingRetentionLease = retentionLeases.get(id);
+        if (existingRetentionLease == null) {
             throw new RetentionLeaseNotFoundException(id);
         }
+        if (retainingSequenceNumber < existingRetentionLease.retainingSequenceNumber()) {
+            throw new IllegalArgumentException(
```

Comment:
We shouldn't hit this for peer-recovery retention leases (we did once and it showed a bug) so at the very least I would like to assert here that the lease in question is not a peer-recovery retention lease. More generally can we assert it's only a CCR lease, or even better, avoid this leniency by changing how CCR followers renew their leases to avoid this kind of situation?

Reply:
Yeah, I agree. But it's hard to guarantee this in CCR. If a shard-follow task fails, then the new task can start with a lower global checkpoint and renew with a lower retaining sequence number.

I added an assertion to make sure that it's not a peer recovery lease in https://github.com/elastic/elasticsearch/pull/46420/commits/405f041b84b0cec4c4711416726836c8e933c5ed.

Reply:
If the new task starts with a lower global checkpoint and attempts to renew the lease backwards as you describe, there is a risk that the extra history it wants to retain has already been discarded. What happens then?

Reply:
We have to perform a remote recovery here.

Reply:
We discussed this today. We acknowledged this leniency when implementing the retention leases for CCR. We might implement a stricter retention lease model in the later versions; but for now, we accept this limitation.